content/en/docs/concepts/configuration/manage-resources-containers.md

```diff
@@ -34,16 +34,20 @@ For example, if you set a `memory` request of 256 MiB for a container, and that
 a Pod scheduled to a Node with 8GiB of memory and no other Pods, then the container can try to use
 more RAM.
 
-If you set a `memory` limit of 4GiB for that container, the kubelet (and
-{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}) enforce the limit.
-The runtime prevents the container from using more than the configured resource limit. For example:
-when a process in the container tries to consume more than the allowed amount of memory,
-the system kernel terminates the process that attempted the allocation, with an out of memory
-(OOM) error.
+Both `CPU` and `memory` limits are applied by the kubelet (and
+{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}),
+and are ultimately enforced by the kernel. On Linux nodes, limits are enforced
+by {{< glossary_tooltip text="cgroups" term_id="cgroup" >}}). However, `CPU` and `memory` limits
+are enforced differently.
```

Comment:
> However, `CPU` and `memory` limits
 are enforced differently.

The above statement is wrong. You already answered correctly in lines 37 - 38, yet contradicted your answer in lines 40 - 41. There's no need to throw the kernel into the conversation here, it ends up confusing the reader.

Only two things are missing from the existing documentation.
1. An outline that the kubelet and container runtime [communicates](https://kubernetes.io/docs/setup/production-environment/container-runtimes/#cgroup-drivers) with **control groups** to enforce the limit.
2. A configured `memory.high` setting in cgroupv2 can prevent a container from getting killed by an OOM error when it exceeds its limit.

Since these are what you are trying to add, I suggest you use the existing documentation and only include what is missing instead of changing all the sentences.

```suggestion
If you set a `memory` limit of 4GiB for that container, the kubelet (and
 {{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}) communicates with control groups to enforce the limit.
 The runtime prevents the container from using more than the configured resource limit. For example:
 when a process in the container tries to consume more than the allowed amount of memory,
 the system kernel terminates the process that attempted the allocation, with an out of memory
```

Reply:
I have clarified this section, I think it's important to mention who is actually doing the enforcing and how. I did mention MemoryQoS but I don't think we should go in depth because we're not moving forward with that effort now.

Reply:
Nice!